fedimint-client/src/transaction/sm.rs

```diff
@@ -114,48 +102,47 @@ impl State for TxSubmissionStates {
     }
 }
 
-impl IntoDynInstance for TxSubmissionStates {
-    type DynType = DynState<DynGlobalClientContext>;
-
-    fn into_dyn(self, instance_id: ModuleInstanceId) -> Self::DynType {
-        DynState::from_typed(instance_id, self)
-    }
-}
-
-async fn trigger_created_rejected(
-    tx: Transaction,
-    context: DynGlobalClientContext,
-) -> Result<String, String> {
-    loop {
-        match context.api().submit_transaction(tx.clone()).await {
-            Ok(submission_result) => {
-                if let Err(submission_error) = submission_result
-                    .try_into_inner(context.decoders())
-                    .map_err(|error| error.to_string())?
+impl TxSubmissionStates {
+    async fn trigger_created_rejected(tx: Transaction, context: DynGlobalClientContext) -> String {
+        loop {
+            match context.api().submit_transaction(tx.clone()).await {
+                Ok(submission_result) => match submission_result.try_into_inner(context.decoders())
                 {
-                    return Ok(submission_error.to_string());
+                    Ok(submission_result) => {
+                        if let Err(transaction_error) = submission_result {
+                            return transaction_error.to_string();
+                        }
+                    }
+                    Err(decode_error) => {
+                        warn!(target: LOG_CLIENT_NET_API, error = %decode_error, "Failed to decode SerdeModuleEncoding")
+                    }
+                },
+                Err(error) => {
+                    error.report_if_important();
                 }
             }
```

Comment:
This is very hard to read, could we maybe use early-return and then document why in some cases we retry while in others we return?

Reply:
I do not see how to factor this better tbh; I have change the variable name submission_result to serve_result and now the formatting is better at least.

Reply:
Tracking which level of error handling we are in is a bit confusing here, maybe it's just the diff view but at first I confused the inner/outer error.